core/src/main/scala/kafka/server/ControllerApis.scala

```diff
@@ -280,25 +281,34 @@ class ControllerApis(val requestChannel: RequestChannel,
       while (iterator.hasNext) {
         val entry = iterator.next()
         val topicName = entry.getValue
+        val topicId = entry.getKey
         if (!authorizedDeleteTopics.contains(topicName)) {
-          // Case 1 or case 4: the topic exists, but we don't have permission to delete it.
-          val topicId = entry.getKey
-          if (topicIdsToResolve.contains(topicId)) {
-            appendResponse(null, topicId, new ApiError(TOPIC_AUTHORIZATION_FAILED))
+          if (authorizedDescribeTopics.contains(topicName)) {
+            if (topicNamesToResolve.contains(topicName)) {
+              // 6. name provided, topic exists, describable => TOPIC_AUTHORIZATION_FAILED
+              appendResponse(topicName, ZERO_UUID, new ApiError(TOPIC_AUTHORIZATION_FAILED))
+            } else {
+              // 2. ID provided, topic present, describeable => TOPIC_AUTHORIZATION_FAILED
+              appendResponse(null, topicId, new ApiError(TOPIC_AUTHORIZATION_FAILED))
+            }
           } else {
-            appendResponse(topicName, ZERO_UUID, new ApiError(TOPIC_AUTHORIZATION_FAILED))
+            if (topicNamesToResolve.contains(topicName)) {
+              // 7. name provided, topic exists, undescribable => UNKNOWN_TOPIC_OR_PARTITION
+              appendResponse(topicName, ZERO_UUID, new ApiError(UNKNOWN_TOPIC_OR_PARTITION))
+            } else {
+              // 3. ID provided, topic present, undescribeable => UNKNOWN_TOPIC_ID
+              appendResponse(null, topicId, new ApiError(UNKNOWN_TOPIC_ID))
+            }
           }
           iterator.remove()
         }
       }
       unknownTopicNameErrors.forEach { (topicName, error) =>
           if (authorizedDescribeTopics.contains(topicName)) {
-            // Case 2: the topic we tried to delete by name doesn't exist, and we have
-            // permission to know that.
+            // 4. name provided, topic missing, undescribable => UNKNOWN_TOPIC_OR_PARTITION
```

Comment:
this should be case 5 (describable) rather than case 4(undescribable ) since it is in `authorizedDescribeTopics.contains(topicName)` rather than `!authorizedDescribeTopics.contains(topicName)`